### PR TITLE
David/twpa opt

### PR DESCRIPTION
**Pull request by DavidSarlle**
This Pr implements the TWPA characterization obtaining the best fidelity while sweeping frequency and power.

image

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [ ] Coverage does not decrease.
- [x] Documentation is updated.
- [x] Compatibility with Qibo modules (Please edit this section if the current pull request is not compatible with the following branches).
    - [x] Qibo: `master`
    - [x] Qibolab: `main`
    - [x] Qibolab_platforms_qrc: `main`

### CODE REVIEW

**Comment by Jacfomg** [comment]
It would be nice to scan the SNR instead of best fidelity, just as we do with the VNA. To avoid this unstable points ypu think were causing iqm to get those results, right ?

**Comment by DavidSarlle** [comment]
@Jacfomg thanks for the comments. 

Finally I have detected that the cause of the decresement in the qubit fidelities is not caused by the TWPA. It is caused by the cross talk between q2 and the rest of the qubits in the iqm chip. When any single qubit gate is made over q0, q1, q3 or q4 the fidelities are good consistently with the current characterization, but when we do sigle qubit gates at the same time, including q2, all the fidelities decrease 20-30%. That results are shown consistenly in all conditions (other chips working, different fridge temp, different hours...) So, I am pretty sure that the problem is not the TWPA. 

Even with that, we have coded this function with the assg. fidelitie as the target to find the best power/freq parameters instead of the SNR. I agree that it could be better, but following the 2 routines that we alredy have in qibocal main, I coded this one. It is good enoygh for the moment and the routine helps alot in the TWPA characterization. So I want to merge it into main asap.

File: src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    frequency_step: float
+    """Frequency step to be probed."""
+    power_width: float
+    """Power total width."""
+    power_step: float
+    """Power step to be probed."""
+    nshots: Optional[int] = None
+    """Number of shots."""
+    relaxation_time: Optional[int] = None
+    """Relaxation time (ns)."""
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py:31]: You can remove those two attributes. They should be imported automatically from `Parameters`

File: src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    with step frequency_step.
+
+    Args:
+        params (:class:`TwpaFrequencyParameters`): input parameters
+        platform (:class:`Platform`): Qibolab's platform
+        qubits (dict): dict of target :class:`Qubit` objects to be characterized
+
+    Returns:
+        data (:class:`TwpaFrequencyData`)
+    """
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py:84]: Outdated docstring

File: src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        for qubit in qubits:
+            platform.qubits[qubit].twpa.local_oscillator.frequency = (
+                initial_twpa_freq[qubit] + freq
+            )
+
+        for power in power_range:
+            for qubit in qubits:
+                platform.qubits[qubit].twpa.local_oscillator.power = (
+                    initial_twpa_power[qubit] + power
+                )
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py:114]: Here I believe you can only 1 for loop around qubits:
```py
for freq in freqs:
    for power in powers:
        set_freq(...)
        set_power(...)
        


```

**Review by andrea-pasquale — COMMENTED** [review]
Thanks @DavidSarlle, I need to test on hardware but the code looks generally good to me.
To avoid decreasing coverage could you please add this routine also in this [runcard](https://github.com/qiboteam/qibocal/blob/main/tests/runcards/protocols.yml)? In this way it will automatically tested using the dummy platform in GitHub.

**Review by andrea-pasquale — COMMENTED** [review]
Thanks @DavidSarlle, I need to test on hardware but the code looks generally good to me.
To avoid decreasing coverage could you please add this routine also in this [runcard](https://github.com/qiboteam/qibocal/blob/main/tests/runcards/protocols.yml)? In this way it will automatically tested using the dummy platform in GitHub.

**Comment by Jacfomg** [comment]
This may be a routine used only to finetune and not very reliable and taking long time for what I realized because it depends on the calibrate qubit states, meaning a higher number of shots than usual to avoid getting lucky on some points let's say. Also, you have no guarantees that this good points will not be just distorting the clouds leaving an unusable qubit. I will strongly avoid touching the TWPA power outside the routine where we get the SNR on the resonator and just finetune the frequency using the 1D routine on most cases. What do you think @andrea-pasquale  and @Edoardo-Pedicillo ?

**Comment by andrea-pasquale** [comment]
> This may be a routine used only to finetune and not very reliable and taking long time for what I realized because it depends on the calibrate qubit states, meaning a higher number of shots than usual to avoid getting lucky on some points let's say. Also, you have no guarantees that this good points will not be just distorting the clouds leaving an unusable qubit. I will strongly avoid touching the TWPA power outside the routine where we get the SNR on the resonator and just finetune the frequency using the 1D routine on most cases. What do you think @andrea-pasquale and @Edoardo-Pedicillo ?

I agree with the fact that it should be used only for finetuning, and the issues that @Jacfomg presented are true.
I saw several plots with 99% fidelity which were mainly the results of a very weird/lucky calibration of the twpa.
Having said this, the major issue I believe is to have protocols to calibrate twpas from scratch using SNR.

We can merge this protocol as it is but we should avoid to run long scans with is, since as we saw it can be very unstable.

**Comment by DavidSarlle** [comment]
> Thanks @DavidSarlle, I need to test on hardware but the code looks generally good to me. To avoid decreasing coverage could you please add this routine also in this [runcard](https://github.com/qiboteam/qibocal/blob/main/tests/runcards/protocols.yml)? In this way it will automatically tested using the dummy platform in GitHub.

Sure. Done. thanks

File: src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py (hunk 1 above)
Comment (DavidSarlle) [reply src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py:31]: REmeoved

File: src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py (hunk 2 above)
Comment (DavidSarlle) [reply src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py:84]: Updated docstrings

File: src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py (hunk 3 above)
Comment (DavidSarlle) [reply src/qibocal/protocols/characterization/readout_optimization/twpa_calibration/frequency_power.py:114]: Done

**Comment by DavidSarlle** [comment]
> > This may be a routine used only to finetune and not very reliable and taking long time for what I realized because it depends on the calibrate qubit states, meaning a higher number of shots than usual to avoid getting lucky on some points let's say. Also, you have no guarantees that this good points will not be just distorting the clouds leaving an unusable qubit. I will strongly avoid touching the TWPA power outside the routine where we get the SNR on the resonator and just finetune the frequency using the 1D routine on most cases. What do you think @andrea-pasquale and @Edoardo-Pedicillo ?
> 
> I agree with the fact that it should be used only for finetuning, and the issues that @Jacfomg presented are true. I saw several plots with 99% fidelity which were mainly the results of a very weird/lucky calibration of the twpa. Having said this, the major issue I believe is to have protocols to calibrate twpas from scratch using SNR.
> 
> We can merge this protocol as it is but we should avoid to run long scans with is, since as we saw it can be very unstable.

Obviously should be use for fine retunning, because it depends on the initial calibration from the lab using the VNA and also depends on the pi pulse calibration. But even with that, for recalibrating all these routines are useful and provide a method to re callibrate without using the VNA and be present in the lab. 

Regarding the duration, as all the routines, depends on the range that you scan. As I said, as you should have an initial callibration from the lab and the VNA, it should not be a large scan

**Comment by Jacfomg** [comment]
> Regardiong the inestability points, I have also seen this points using the twpa freq or power. But with the 2d scan is much easier to identify this regions and be sure that you are in the correct area for a given power and freq.

The main problem is that we would be trying to measure these unstable points relying on the results of another unstable routine, the calibrate qubits states, so this be way trickier. I would say that the first step for the finetune would be the resonator SNR with power and frequency on the TWPA.

**Comment by DavidSarlle** [comment]
@andrea-pasquale can we merge this PR. Until we do not implement other methods like the resonator spec vs TWPA freq/power and fit the best SNR point, this routine is still useful for the people that is re-calibrating as we can see what are the unestable points and we can refine the TWPA once the pi pulses are defined.

**Review by Edoardo-Pedicillo — APPROVED** [review]
Thank @DavidSarlle LGTM

**Review by Jacfomg — APPROVED** [review]
I hope approving this routine will mean that we won't get the others, but still, you are calibrating using an unstable figure of merit, so if you are going to use try to have a big number of shots above 10_000 to 50_000 at least to get reproducible results and not just some lucky blobs that may misslead the calibration.

**Review by andrea-pasquale — APPROVED** [review]

**Pull request merged**